Provide more clues to what's causing the error SSL: WRONG_VERSION_NUMBER

<!--

*******************************************************************************

Please ask questions in our forum at 

https://discuss.elastic.co/c/elasticsearch/rally

GitHub is reserved for bug reports and feature requests only and we will 
close your issue immediately if it is just a question.
 
*******************************************************************************

If you are in fact posting a bug report or a feature request, please include 
one and only one of the below blocks in your new issue.

-->

<!--
If you are filing a bug report, please remove the below feature
request block and provide responses for all of the below items.
-->

**Rally version** (get with `esrally --version`): 1.2.1

**Invoked command**: 

`esrally --offline --target-hosts=https://<my_ip>:9200 --pipeline=benchmark-only --client-options="timeout:120,use_ssl:true,verify_certs --report-file=test-1.csv --track-path /mnt/rally/benchmarks/tracks/default/http_logs/`

**OS version**:

```
cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
DISTRIB_DESCRIPTION="Ubuntu 19.04"
```

**Description of the problem including expected versus actual behavior**:

When using SSL options against a non-ssl endpoint, the error message is a bit cryptic as to what has failed.

```
[ERROR] Cannot race. Error in race control (ConnectionError([SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1056)) caused by: SSLError([SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1056)))
```

**Steps to reproduce**:

I saw this issue using `--client-options="timeout:120,use_ssl:true,verify_certs:false` against an endpoint with  `--target-hosts=https://whateverhost` 

Setting `use_ssl` to false and leaving verify_certs:false still saw the error occurring.
I had to remove the https and use http instead to have it correctly connect to the endpoint.

**Provide logs (if relevant)**:

**Describe the feature**:

A list of errors with possible causes/resolutions in a troubleshooting guide within the Rally docs is probably enough to steer users to the fix as I don't think this would be all that common.

Additionally, a more informative error message might give users the right hint faster (`"is this a https endpoint?"`)

I could not reproduce this exactly in the way you describe it because Rally first failed when checking the distribution-version so I had to specify it manually:

```
esrally --test-mode --pipeline=benchmark-only --target-host=https://127.0.0.1:9200 --distribution-version=7.3.0 --client-options="timeout:120,use_ssl:true,verify_certs:false"  --on-error=abort
```

After that it manifested as an error message on the command line that the REST API is not available:

```
[ERROR] Cannot race. Could not launch cluster
	Elasticsearch REST API layer is not available. Forcefully terminated cluster.
```

But in the logs I could see the error message you're refering to.

I'll soon have a fix ready that instead will show this:

```
[ERROR] Cannot race. Could not launch cluster
	Could not connect to cluster via https. Is this a https endpoint?
		ConnectionError([SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:719)) caused by: SSLError([SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:719))
```

So it provides actionable advice but also the root cause in the error message.

--- a/esrally/mechanic/launcher.py
+++ b/esrally/mechanic/launcher.py
@@ -25,7 +25,7 @@ from time import monotonic as _time
 
 from esrally import config, time, exceptions, client
 from esrally.mechanic import telemetry, cluster, java_resolver
-from esrally.utils import process, jvm
+from esrally.utils import process
 
 
 def wait_for_rest_layer(es, max_attempts=20):
@@ -34,8 +34,13 @@ def wait_for_rest_layer(es, max_attempts=20):
         try:
             es.info()
             return True
+        except elasticsearch.ConnectionError as e:
+            if "SSL: UNKNOWN_PROTOCOL" in str(e):
+                raise exceptions.LaunchError("Could not connect to cluster via https. Is this a https endpoint?", e)
+            else:
+                time.sleep(1)
         except elasticsearch.TransportError as e:
-            if e.status_code == 503 or isinstance(e, elasticsearch.ConnectionError):
+            if e.status_code == 503:
                 time.sleep(1)
             elif e.status_code == 401:
                 time.sleep(1)

--- a/tests/mechanic/launcher_test.py
+++ b/tests/mechanic/launcher_test.py
@@ -14,7 +14,6 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-import tempfile
 from datetime import datetime
 import io, os
 import uuid
@@ -78,7 +77,7 @@ class MockClient:
     def info(self):
         if self.client_options.get("raise-error-on-info", False):
             import elasticsearch
-            raise elasticsearch.ConnectionError("Unittest error")
+            raise elasticsearch.TransportError(401, "Unauthorized")
         return self._info
 
     def search(self, *args, **kwargs):
@@ -314,3 +313,49 @@ class ClusterLauncherTests(TestCase):
         with self.assertRaisesRegex(exceptions.LaunchError,
                                     "Elasticsearch REST API layer is not available. Forcefully terminated cluster."):
             cluster_launcher.start()
+
+
+class RestLayerTests(TestCase):
+    @mock.patch("elasticsearch.Elasticsearch", autospec=True)
+    def test_successfully_waits_for_rest_layer(self, es):
+        self.assertTrue(launcher.wait_for_rest_layer(es, max_attempts=3))
+
+    # don't sleep in realtime
+    @mock.patch("time.sleep")
+    @mock.patch("elasticsearch.Elasticsearch", autospec=True)
+    def test_retries_on_transport_errors(self, es, sleep):
+        import elasticsearch
+
+        es.info.side_effect = [
+            elasticsearch.TransportError(503, "Service Unavailable"),
+            elasticsearch.TransportError(401, "Unauthorized"),
+            {
+                "version": {
+                    "number": "5.0.0",
+                    "build_hash": "abc123"
+                }
+            }
+        ]
+        self.assertTrue(launcher.wait_for_rest_layer(es, max_attempts=3))
+
+    # don't sleep in realtime
+    @mock.patch("time.sleep")
+    @mock.patch("elasticsearch.Elasticsearch", autospec=True)
+    def test_dont_retries_eternally_on_transport_errors(self, es, sleep):
+        import elasticsearch
+
+        es.info.side_effect = elasticsearch.TransportError(401, "Unauthorized")
+        self.assertFalse(launcher.wait_for_rest_layer(es, max_attempts=3))
+
+    @mock.patch("elasticsearch.Elasticsearch", autospec=True)
+    def test_ssl_error(self, es):
+        import elasticsearch
+        import urllib3.exceptions
+
+        es.info.side_effect = elasticsearch.ConnectionError("N/A",
+                                                            "[SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:719)",
+                                                            urllib3.exceptions.SSLError(
+                                                                "[SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:719)"))
+        with self.assertRaisesRegex(expected_exception=exceptions.LaunchError,
+                                    expected_regex="Could not connect to cluster via https. Is this a https endpoint?"):
+            launcher.wait_for_rest_layer(es, max_attempts=3)

